src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -25,27 +25,52 @@ internal sealed class ODataUtf8JsonWriter : IJsonWriter, IDisposable
         private readonly float bufferFlushThreshold;
 
         private readonly Stream outputStream;
+        private readonly Stream writeStream;
         private readonly Utf8JsonWriter writer;
         private readonly int bufferSize;
         private readonly bool isIeee754Compatible;
         private readonly bool leaveStreamOpen;
-        private readonly Encoding encoding = Encoding.UTF8;
+        private readonly Encoding utf8Encoding = Encoding.UTF8;
+        private readonly Encoding outputEncoding;
         private bool disposed;
 
-        public ODataUtf8JsonWriter(Stream outputStream, bool isIeee754Compatible, int bufferSize = DefaultBufferSize, bool leaveStreamOpen = false)
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUtf8JsonWriter"/>.
+        /// </summary>
+        /// <param name="outputStream">The stream to write to.</param>
+        /// <param name="isIeee754Compatible">When true, longs and decimals will be written out as strings.</param>
+        /// <param name="encoding">The text encoding to write to.</param>
+        /// <param name="bufferSize">The internal buffer size.</param>
+        /// <param name="leaveStreamOpen">Whether to leave the <paramref name="outputStream"/> open when this writer gets disposed.</param>
+        public ODataUtf8JsonWriter(Stream outputStream, bool isIeee754Compatible, Encoding encoding, int bufferSize = DefaultBufferSize, bool leaveStreamOpen = false)
         {
             Debug.Assert(outputStream != null, "outputStream != null");
             this.outputStream = outputStream;
             this.isIeee754Compatible = isIeee754Compatible;
             this.bufferSize = bufferSize;
             // flush when we're close to the buffer capacity to avoid allocating bigger buffers
             this.bufferFlushThreshold = 0.9f * this.bufferSize;
+            this.leaveStreamOpen = leaveStreamOpen;
+            this.outputEncoding = encoding;
+
+            if (this.outputEncoding.CodePage == Encoding.UTF8.CodePage)
+            {
+                this.writeStream = this.outputStream;

```

Comment:
Why do we need a second field for this? Can't we just have `outputStream` and set it to a `TranscodingWriteStream` if necessary?

Reply:
In fact, it seems like instantiating the `TranscodingWriteStream` can be done by the caller, and they pass it to our constructor. We should probably go that route, I would think.

Reply:
Regarding why we need both `writeStream` and `outputStream` fields? I'll check the code again, maybe we don't need both. I think we always dispose `writeStream` if it's different from `outputStream`. And `outputStream` is optionally disposed.